src/Microsoft.Health.Fhir.Core/Features/Search/Expressions/Parsers/SearchParameterExpressionParser.cs

```diff
@@ -84,54 +84,61 @@ public class SearchParameterExpressionParser : ISearchParameterExpressionParser
-                            string.Format(CultureInfo.InvariantCulture, Core.Resources.NumberOfCompositeComponentsExceeded, searchParameter.Name));
-                    }
+                        IReadOnlyList<string> compositeValueParts = orParts[orIndex].SplitByCompositeSeparator();
 
-                    var compositeExpressions = new Expression[compositeValueParts.Count];
+                        if (compositeValueParts.Count > searchParameter.Component.Count)
+                        {
+                            throw new InvalidSearchOperationException(
+                                string.Format(CultureInfo.InvariantCulture, Resources.NumberOfCompositeComponentsExceeded, searchParameter.Name));
+                        }
 
-                    var searchParameterComponentInfos = searchParameter.Component.ToList();
+                        var compositeExpressions = new Expression[compositeValueParts.Count];
 
-                    for (int i = 0; i < compositeValueParts.Count; i++)
-                    {
-                        var component = searchParameterComponentInfos[i];
+                        var searchParameterComponentInfos = searchParameter.Component.ToList();
+
+                        for (int componentIndex = 0; componentIndex < compositeValueParts.Count; componentIndex++)
+                        {
+                            SearchParameterComponentInfo component = searchParameterComponentInfos[componentIndex];
+
+                            // Find the corresponding search parameter info.
+                            SearchParameterInfo componentSearchParameter = _searchParameterDefinitionManager.GetSearchParameter(component.DefinitionUrl);
 
-                        // Find the corresponding search parameter info.
-                        SearchParameterInfo componentSearchParameter = _searchParameterDefinitionManager.GetSearchParameter(component.DefinitionUrl);
+                            string componentValue = compositeValueParts[componentIndex];
 
-                        string componentValue = compositeValueParts[i];
+                            compositeExpressions[componentIndex] = Build(
+                                componentSearchParameter,
+                                modifier: null,
+                                componentIndex: componentIndex,
+                                value: componentValue);
+                        }
 
-                        compositeExpressions[i] = Build(
-                            componentSearchParameter,
-                            modifier: null,
-                            componentIndex: i,
-                            value: componentValue);
+                        orExpressions[orIndex] = Expression.And(compositeExpressions);
                     }
 
-                    outputExpression = Expression.And(compositeExpressions);
+                    outputExpression = orExpressions.Length == 1 ? orExpressions[0] : Expression.Or(orExpressions);
```

Comment:
Any tests we can add here? https://github.com/microsoft/fhir-server/blob/master/src/Microsoft.Health.Fhir.Shared.Core.UnitTests/Features/Search/Expressions/Parsers/ExpressionParserTests.cs

Reply:
I've run that tests, and none of them actually hitting parser method, so I'm not sure what exactly they test...